---
title: "WGCNA_Cvirginica_Feeding"
author: "Samuel Gurr"
date: "2023-08-24"
output: html_document
---

NOTE: the adjacency and TOM matrix steps have long computation times
used save(<list files>, file = "filename.RData") to save at steps along the way
load steps below and skip to relevent chunks to analyze data

# SETUP
```{r setup, include= FALSE}
# LOAD PACKAGES
library(WGCNA) # note: this was previously installed with the command `BiocManager::install("WGCNA")`
library(dplyr)
library(zoo)
library(DESeq2)
library(ComplexHeatmap)
library(circlize)
library(reshape)
library(ggplot2)
library(hrbrthemes)

# SET WORKING DIRECTORY

knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Cvirginica_KM_paper/RAnalysis") # sets the working 

```


# LOAD THE PRE-FILTERED ABUNDANCE MATRIX
```{r}
# read txt file gene count matrix
Filt_abund  <- read.csv(file="Output/Filtering_reads/All.filtered_5cpm50perc.csv", sep=',', header=TRUE) 
dim(Filt_abund) # 12639    25 - 25 columns and 38977 total genes 

# The following setting is important, do not omit. (as recommended by WGCNA authors - view tutorial)
options(stringsAsFactors = FALSE)

```


# PREPROCESSING THE METADATA  
```{r}
#  Read here for the pre processing steps using WGCNA!
#  https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/1471-2105-9-559.pdf


# ===================================================================================
#
# Prepare abundance matrix
# ===================================================================================


# read count data  ========================================================== #
Filt_abund_matrix <- data.frame(Filt_abund[,-1], row.names=Filt_abund[,1]) 
dim(Filt_abund_matrix) # 38977    24 - 24 samples, gene IDs now row names



# ===================================================================================
#
# Prepare Trait data (Phys and Treatment groups)
# ===================================================================================

# Treatment data

Metadata <- read.csv(file="Data/SummaryforWGCNAtreat.csv", sep=',', header=TRUE)
dim(Metadata) # 24 samples, 11 columns

# ABOUT: here we have the following: 
# experiment day, food ration, wild v. non-wild, and quantitative phys data
# all but the quant phys data all bins to group WGCNa analysis - lets do this for the phys! 

# reformat phys data as numeric string bins ================================== # 

# round down, accoutn for sig figs
signif.floor <- function(x, n){
  pow <- floor(log10(abs(x))) + 1 - n
  y   <- floor(x / 10 ^ pow) * 10^pow#handle the x = 0 case 
  y[x==0] <- 0
  y
}
# round down, accoutn for sig figs
signif.ceiling <- function(x, n){
  pow <- ceiling(log10(abs(x))) + 1 - n
  y   <- ceiling(x / 10 ^ pow) * 10^pow#handle the x = 0 case 
  y[x==0] <- 0
  y
}


# for loop through and make bins 
# 
for (i in (1:ncol(Metadata[,c(6:11)]))) {
  # looped string for new column names
  colname_loop <- colnames(Metadata[i+5])
  # our bins are defined from the number of standard deviations 
  # between the min and max values 
  min          <- signif.floor(min(Metadata[i+5]),3) 
  max          <- signif.ceiling(max(Metadata[i+5]),3) 
  sd           <- sd(as.numeric(unlist(Metadata[i+5])))
  bin_num      <- round((max-min)/sd) # #of bins
  # loop new column as  bins using dplyr 'ntile' - use as.character for WGCNA
  Metadata[[paste0(colname_loop,'_bins')]] <- as.character(dplyr::ntile(Metadata[i+5], n = bin_num))
}

# IMPORTANT FOR FUTURE REFERENCE IN ROW BELOW
# View(Metadata) # here is is, refer to the values of these bins later 


Metadata_trim       <- Metadata[,-c(2,6:11)] # removed the raw quantitative data, only use the categorical bins
Metadata_trim$Group <- paste0('Day',Metadata_trim$Day,'_',
                              Metadata_trim$Line, '_', 
                              Metadata_trim$Treatment)
# Div data by time ==================================================== #

# metadata 
# all metdata as Metadata
Metadata_D0 <- Metadata_trim %>%  dplyr::filter(Day == 0) # metadata for time 0
Metadata_D1 <- Metadata_trim %>%  dplyr::filter(Day == 1) # metadata for time 1
# abundance matrix
Filt_abund_matrix_D0 <- Filt_abund_matrix %>% dplyr::select(Metadata_D0$Sample.ID) # select sample IDs for day 0
Filt_abund_matrix_D1 <- Filt_abund_matrix %>% dplyr::select(Metadata_D1$Sample.ID) # select sample IDs for day 0

# sanity check ======================================================== # 
dim(Filt_abund_matrix)[2]    ==  dim(Metadata)[1]# TRUE - each contains all 24 samples (ALL DATA)
dim(Filt_abund_matrix_D0)[2] ==  dim(Metadata_D0)[1]# TRUE - each contains all 12 samples (Day 0 data)
dim(Filt_abund_matrix_D1)[2] ==  dim(Metadata_D1)[1]# TRUE - each contains all 12 samples (Day 1 data)


```


# DESEQdataset 
```{r}
# ===================================================================================
#
# DESeqDataSet or 'dds' object (using DESeq2) 
# vst transform the 'dds' onject for WGCNA 
# 
# ===================================================================================

# create dds  ========================================================== #

# NOTE: ~1 stands for no design; user will need to add a design for differential testing
# however for our purpose of just creating an object to transform, we do not need a design here...

# all
dds_all <- DESeqDataSetFromMatrix(countData = Filt_abund_matrix,
                                  colData = Metadata, design = ~ 1) # DESeq Data Set (dds)
dds_all # view the DESeqDataSet - notice the colData containg our critical treatment and sample ID data, rownames, etc. 

# day 0
dds_D0 <- DESeqDataSetFromMatrix(countData = Filt_abund_matrix_D0,
                                  colData = Metadata_D0, design = ~ 1) # DESeq Data Set (dds)
dds_D0 # view the DESeqDataSet - notice the colData containg our critical treatment and sample ID data, rownames, etc. 

# day 1
dds_D1 <- DESeqDataSetFromMatrix(countData = Filt_abund_matrix_D1,
                                  colData = Metadata_D1, design = ~ 1) # DESeq Data Set (dds)
dds_D1 # view the DESeqDataSet - notice the colData containg our critical treatment and sample ID data, rownames, etc. 


```


# TRANSFORM & TRANSPOSE THE DESEQdataset 
```{r}
# transform the data  ========================================================== #

# all
dds_all_vst    <- vst(dds_all) # transform it vst (variance stabilized transformation)
dds_all_vst    <- assay(dds_all) # call only the transformed coutns in the dds object
dds_all_vst_t  <- t(dds_all_vst) # transpose columns to rows and vice versa
dim(dds_all_vst_t) #  24 12639
# day 0
dds_D0_vst   <- vst(dds_D0) # transform it vst (variance stabilized transformation)
dds_D0_vst   <- assay(dds_D0_vst) # call only the transformed coutns in the dds object
dds_D0_vst_t <- t(dds_D0_vst) # transpose columns to rows and vice versa
dim(dds_D0_vst_t) #  12 12639
# day 1
dds_D1_vst   <- vst(dds_D1) # transform it vst (variance stabilized transformation)
dds_D1_vst   <- assay(dds_D1_vst) # call only the transformed coutns in the dds object
dds_D1_vst_t <- t(dds_D1_vst) # transpose columns to rows and vice versa
dim(dds_D1_vst_t) # 12 12639
```


# SAMPLETREE TIME - (omission of samples possible here)
```{r}
# =================================================================================== #
#
# WGCNA Sample tree - omit outlier sample(s)
#
# =================================================================================== 3

# checks before we start....

# determine outlier(s) with a sample tree    ====================================== #
# make a tree with transposed data, view and add cut threshold
# integrate cuts to data - omitting outlier samples (if applicable!)


# all  ======================================== #

dim(dds_all_vst_t) #  12639 genes; 24  samples
gsg_all = goodSamplesGenes(dds_all_vst_t, verbose = 3);  # We first check for genes and samples with too many missing values:
gsg_all$allOK # If the statement returns TRUE, all genes have passed the cuts. 

sampleTree_all = hclust(dist(dds_all_vst_t), method = "average") # Next we cluster the samples to see if there are any obvious outliers.
sizeGrWindow(12,9) # The user should change the dimensions if the window is too large or too small.
par(cex = 0.6);
par(mar = c(0,4,2,0))
png("Output/All/All_ClusterTree_Precut.png", 1000, 1000, pointsize=20)
plot(sampleTree_all, main = "Sample clustering to detect outliers", 
     sub="", 
     xlab="", 
     cex.lab = 1.5, 
     cex.axis = 1.5, 
     cex.main = 2) # appears there is one outlier!; can remove by hand or an automatic appraoch 
abline(h = 60000, col = "red") # add line to plot to show the cut-off 
dev.off() # one outlier 

clust_all = cutreeStatic(sampleTree_all, cutHeight = 60000, minSize = 10) # Determine cluster under the line
table(clust_all) # 0 = cut; 1 = kept; says it will cut 1 and save 23; exactly what we want!  
keepSamples_all = (clust_all==1) # 'keepsamples' boolean to call the main dataset - notice there are TWO occurrences of FALSE - these are C6.larva and B12.larva
dds_all_vst_t_trim = (dds_all_vst_t[keepSamples_all, ]) # integrate the boolean 'keepsamples' to omit outliers determined in the sample tree above 
sampleTree_all_trim = hclust(dist(dds_all_vst_t_trim), method = "average") # Next we cluster the samples to see if there are any obvious outliers.





nGenes = ncol(dds_all_vst_t_trim) # number of samples == 12639 - the cut tree removed 2 samples 
nSamples = nrow(dds_all_vst_t_trim) # number of genes == 23  
colnames(t(dds_all_vst_t_trim))
dim(Metadata_trim) # 24 samples - not yet cut! = nte that Metada_trim has the binned columns 'trimmed' of raw phys data

Metadata_all <- Metadata_trim %>% dplyr::filter(Sample.ID %in% colnames(t(dds_all_vst_t_trim)))
dim(Metadata_all) # 23 17 - we have 23 sames now, ddone! - use Metadata_all for all data references! 






# day 0 ======================================== #

dim(dds_D0_vst_t) #  12 12639
gsg_D0 = goodSamplesGenes(dds_D0_vst_t, verbose = 3);  # We first check for genes and samples with too many missing values:
gsg_D0$allOK # If the statement returns TRUE, all genes have passed the cuts. 


sampleTree_D0 = hclust(dist(dds_D0_vst_t), method = "average") # Next we cluster the samples (in contrast to clustering genes that will come later)  to see if there are any obvious outliers.
sizeGrWindow(12,9) # The user should change the dimensions if the window is too large or too small.
par(cex = 0.6);
par(mar = c(0,4,2,0))
png("Output/Day0/Day0_ClusterTree_Precut.png", 1000, 1000, pointsize=20)
plot(sampleTree_D0, main = "Sample clustering to detect outliers", 
     sub="", 
     xlab="", 
     cex.lab = 1.5, 
     cex.axis = 1.5, 
     cex.main = 2) # appears there is one outlier!; can remove by hand or an automatic appraoch 
dev.off() # one outlier 

# no ouliers detected!


# day 1  ======================================== #

dim(dds_D1_vst_t) #  38977 genes; 12  samples
gsg_D1 = goodSamplesGenes(dds_D1_vst_t, verbose = 3);  # We first check for genes and samples with too many missing values:
gsg_D1$allOK # If the statement returns TRUE, all genes have passed the cuts. 


sampleTree_D1 = hclust(dist(dds_D1_vst_t), method = "average") # Next we cluster the samples (in contrast to clustering genes that will come later)  to see if there are any obvious outliers.
sizeGrWindow(12,9) # The user should change the dimensions if the window is too large or too small.
par(cex = 0.6);
par(mar = c(0,4,2,0))
png("Output/Day1/Day1_ClusterTree_Precut.png", 1000, 1000, pointsize=20)
plot(sampleTree_D1, main = "Sample clustering to detect outliers", 
     sub="", 
     xlab="", 
     cex.lab = 1.5, 
     cex.axis = 1.5, 
     cex.main = 2) # appears there is one outlier!; can remove by hand or an automatic appraoch 
dev.off() # one outlier 

# no ouliers detected!


```


# SAMPLES OVERLAID ON SAMPLETREE (prelim views are fun)
```{r}
# ===================================================================================
#
# cluster samples by Trait
# ===================================================================================


# format treatment data as numeric string bins 


# 'Treatment' fed v. starved  ================== #

# all
All_Treatment  <-  Metadata_all %>% dplyr::select('Treatment') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(fed = as.factor(as.numeric(Treatment == "fed")))  %>%  # call occurrence of 'A' as 0s and 1s (factor)
  dplyr::mutate(starved = as.factor(as.numeric(Treatment == "starved"))) %>%  # call occurrence of 'M'  as 0s and 1s (factor)
  dplyr::select(-Treatment)
All_Treatment  # final dataset of 0,1 for treatment groups - temperature only!
png("Output/All/Feeding_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_treat = labels2colors(All_Treatment); # Convert traits to a color representation:
plotDendroAndColors(sampleTree_all_trim, traitColors_treat, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(All_Treatment), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

# day 0
D0_Treatment  <-  Metadata_D0 %>% dplyr::select('Treatment') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(fed = as.factor(as.numeric(Treatment == "fed")))  %>%  # call occurrence of 'A' as 0s and 1s (factor)
  dplyr::mutate(starved = as.factor(as.numeric(Treatment == "starved"))) %>%  # call occurrence of 'M'  as 0s and 1s (factor)
  dplyr::select(-Treatment)
D0_Treatment  # final dataset of 0,1 for treatment groups - temperature only!
png("Output/Day0/Feeding_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_treat = labels2colors(D0_Treatment); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D0, traitColors_treat, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D0_Treatment), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

# day 1
D1_Treatment  <-  Metadata_D1 %>% dplyr::select('Treatment') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(fed = as.factor(as.numeric(Treatment == "fed")))  %>%  # call occurrence of 'A' as 0s and 1s (factor)
  dplyr::mutate(starved = as.factor(as.numeric(Treatment == "starved"))) %>%  # call occurrence of 'M'  as 0s and 1s (factor)
  dplyr::select(-Treatment)
D1_Treatment  # final dataset of 0,1 for treatment groups - temperature only!

png("Output/Day1/Feeding_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_treat = labels2colors(D1_Treatment); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D1, traitColors_treat, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D1_Treatment), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()


# 'Line'  ========================= #

#all
All_Line  <-  Metadata_all %>% dplyr::select('Line') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate(lola = as.factor(as.numeric(Line == "lola")))  %>%  
  dplyr::mutate(wild = as.factor(as.numeric(Line == "wild"))) %>%  
  dplyr::select(-Line)
All_Line  # final dataset of 0,1 for Line groups - temperature only!
png("Output/All/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_line = labels2colors(All_Line); # Convert traits to a color representation
plotDendroAndColors(sampleTree_all_trim, traitColors_line, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(All_Line), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

D0_Line  <-  Metadata_D0 %>% dplyr::select('Line') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate(lola = as.factor(as.numeric(Line == "lola")))  %>% 
  dplyr::mutate(wild = as.factor(as.numeric(Line == "wild"))) %>%  
  dplyr::select(-Line)
D0_Line  # final dataset of 0,1 for Line groups - temperature only!
png("Output/Day0/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_line = labels2colors(D0_Line); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D0, traitColors_line, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D0_Line), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

D1_Line  <-  Metadata_D1 %>% dplyr::select('Line') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate(lola = as.factor(as.numeric(Line == "lola")))  %>% 
  dplyr::mutate(wild = as.factor(as.numeric(Line == "wild"))) %>%  
  dplyr::select(-Line)
D1_Line  # final dataset of 0,1 for Line groups - temperature only!
png("Output/Day1/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_line = labels2colors(D1_Line); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D1, traitColors_line, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D1_Line), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()




# all treatment groups =================== #

# all
All_Group         <- Metadata_all %>% 
  dplyr::select('Group') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(D0_L_fed     = as.factor(as.numeric(Group == "Day0_lola_fed")))  %>%  
  dplyr::mutate(D0_L_starved = as.factor(as.numeric(Group == "Day0_lola_starved")))  %>%  
  dplyr::mutate(D0_W_fed     = as.factor(as.numeric(Group == "Day0_wild_fed")))  %>%  
  dplyr::mutate(D0_W_starved = as.factor(as.numeric(Group == "Day0_wild_starved")))  %>%   
  dplyr::mutate(D1_L_fed     = as.factor(as.numeric(Group == "Day1_lola_fed")))  %>% 
  dplyr::mutate(D1_L_starved = as.factor(as.numeric(Group == "Day1_lola_starved")))  %>%  
  dplyr::mutate(D1_W_fed     = as.factor(as.numeric(Group == "Day1_wild_fed")))  %>%  
  dplyr::mutate(D1_W_starved = as.factor(as.numeric(Group == "Day1_wild_starved")))  %>% 
  dplyr::select(-Group)
All_Group
png("Output/All/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_group = labels2colors(All_Group); # Convert traits to a color representation
plotDendroAndColors(sampleTree_all_trim, traitColors_group, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(All_Group), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

# day 0
D0_Group         <- Metadata_D0 %>% 
  dplyr::select('Group') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(D0_L_fed     = as.factor(as.numeric(Group == "Day0_lola_fed")))  %>%  
  dplyr::mutate(D0_L_starved = as.factor(as.numeric(Group == "Day0_lola_starved")))  %>%  
  dplyr::mutate(D0_W_fed     = as.factor(as.numeric(Group == "Day0_wild_fed")))  %>%  
  dplyr::mutate(D0_W_starved = as.factor(as.numeric(Group == "Day0_wild_starved")))  %>%  
  dplyr::select(-Group)
D0_Group
png("Output/Day0/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_group = labels2colors(D0_Group); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D0, traitColors_group, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D0_Group), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

# day1
D1_Group         <- Metadata_D1 %>% 
  dplyr::select('Group') %>% # primary treatment as Ambient (A) vs. Moderate (M)
  dplyr::mutate(D1_L_fed     = as.factor(as.numeric(Group == "Day1_lola_fed")))  %>%  
  dplyr::mutate(D1_L_starved = as.factor(as.numeric(Group == "Day1_lola_starved")))  %>%  
  dplyr::mutate(D1_W_fed     = as.factor(as.numeric(Group == "Day1_wild_fed")))  %>% 
  dplyr::mutate(D1_W_starved = as.factor(as.numeric(Group == "Day1_wild_starved")))  %>%  
  dplyr::select(-Group)
D1_Group
png("Output/Day1/Line_ClusterTree_Temperature.png", 1000, 1000, pointsize=20)
traitColors_group = labels2colors(D1_Group); # Convert traits to a color representation
plotDendroAndColors(sampleTree_D1, traitColors_group, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(D1_Group), 
                    main = "Sample dendrogram and trait heatmap (temperature)")
dev.off()

```

# IF YOU COMPLETED THE SOFT THRESHOLDS, ADJACENCY, TOM, MEs, -- SKIP TO  LATER CHECKPOINT!

# FIND SOFT THRESHOLDS
```{r}
# ===================================================================================
#
# soft threshold 
#
# ===================================================================================
# ABOUT: pickSoftThreshold 
#  performs the analysis of network topology and aids the
#  user in choosing a proper soft-thresholding power.
#  The user chooses a set of candidate powers (the function provides suitable default values)
#  function returns a set of network indices that should be inspected


# FIGURES OUTPUT: The left panel... shows the scale-free fit index (y-axis) 
# as a function of the soft-thresholding power (x-axis). 
# We choose the power 9, which is the lowest power for which the scale-free 
# topology index curve attens out upon reaching a high value (in this case, roughly 0.90).
# The right panel.... displays the mean connectivity
# (degree, y-axis) as a function of the soft-thresholding power (x-axis).


# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to=20, by=2))


# all ================================================== #

# Call the network topology analysis function
sft_all = pickSoftThreshold(dds_all_vst_t_trim, powerVector = powers, verbose = 5) #...wait for this to finish

# png to output 
sizeGrWindow(9, 5) # set window size 
png("Output/All/All_ScaleTopology_SoftThresh.png", 1000, 1000, pointsize=20)
        par(mfrow = c(1,2));
        cex1 = 0.9;
        plot(sft_all$fitIndices[,1], -sign(sft_all$fitIndices[,3])*sft_all$fitIndices[,2], # Scale-free topology fit index as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
             main = paste("Scale independence"));
        text(sft_all$fitIndices[,1], -sign(sft_all$fitIndices[,3])*sft_all$fitIndices[,2],
             labels=powers,cex=cex1,col="red");
        abline(h=0.90,col="red") # look at at cut off at power of 3 - this line corresponds to using an R^2 cut-off of h
        
        plot(sft_all$fitIndices[,1], sft_all$fitIndices[,5], # Mean connectivity as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
             main = paste("Mean connectivity"))
        text(sft_all$fitIndices[,1], sft_all$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off() # output 

# sft threshold for 'All' == 8



# day 0 ================================================== #


# call the network topology analysis function
sft_d0 = pickSoftThreshold(dds_D0_vst_t, powerVector = powers, verbose = 5) #...wait for this to finish

# png to output 
sizeGrWindow(9, 5) # set window size 
png("Output/Day0/d0_ScaleTopology_SoftThresh.png", 1000, 1000, pointsize=20)
        par(mfrow = c(1,2));
        cex1 = 0.9;
        plot(sft_d0$fitIndices[,1], -sign(sft_d0$fitIndices[,3])*sft_d0$fitIndices[,2], # Scale-free topology fit index as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
             main = paste("Scale independence"));
        text(sft_d0$fitIndices[,1], -sign(sft_d0$fitIndices[,3])*sft_d0$fitIndices[,2],
             labels=powers,cex=cex1,col="red");
        abline(h=0.90,col="red") # look at at cut off at power of 3 - this line corresponds to using an R^2 cut-off of h
        
        plot(sft_d0$fitIndices[,1], sft_d0$fitIndices[,5], # Mean connectivity as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
             main = paste("Mean connectivity"))
        text(sft_d0$fitIndices[,1], sft_d0$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off() # output 

# sft threshold for 'Day 0' == 5


# day 1 ================================================== #


# call the network topology analysis function
sft_d1 = pickSoftThreshold(dds_D1_vst_t, powerVector = powers, verbose = 5) #...wait for this to finish

# png to output 
sizeGrWindow(9, 5) # set window size 
png("Output/Day1/d1_ScaleTopology_SoftThresh.png", 1000, 1000, pointsize=20)
        par(mfrow = c(1,2));
        cex1 = 0.9;
        plot(sft_d1$fitIndices[,1], -sign(sft_d1$fitIndices[,3])*sft_d1$fitIndices[,2], # Scale-free topology fit index as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
             main = paste("Scale independence"));
        text(sft_d1$fitIndices[,1], -sign(sft_d1$fitIndices[,3])*sft_d1$fitIndices[,2],
             labels=powers,cex=cex1,col="red");
        abline(h=0.90,col="red") # look at at cut off at power of 3 - this line corresponds to using an R^2 cut-off of h
        
        plot(sft_d1$fitIndices[,1], sft_d1$fitIndices[,5], # Mean connectivity as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
             main = paste("Mean connectivity"))
        text(sft_d1$fitIndices[,1], sft_d1$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off() # output 

# sft threshold for 'Day 1' == 8
```


# WGCNA MODULE CONSTRUCTION (finally, we hath arrived young padawn)

## Adjacency matrices
```{r}

#=====================================================================================
#
# Start the step-wise module construction:  
# Step 1 = create adjacency matrix 
# https://peterlangfelder.com/2018/11/25/signed-or-unsigned-which-network-type-is-preferable/
# https://www.rdocumentation.org/packages/WGCNA/10cpm/versions/1.69/topics/adjacency
# https://ramellose.github.io/networktutorials/wgcna.html
# https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/10cpm/TechnicalReports/signedTOM.pdf
#=====================================================================================
softPower_all = 8 # set your soft threshold based on the plots above 
softPower_d0  = 5 # set your soft threshold based on the plots above 
softPower_d1  = 8 # set your soft threshold based on the plots above 

# signed - must call te type, defaults to unsigned
adjacency_sign_all = adjacency(dds_all_vst_t_trim, power = softPower_all, type="signed") # this takes a long time.. just wait...
adjacency_sign_d0  = adjacency(dds_D0_vst_t, power = softPower_d0, type="signed") # this takes a long time.. just wait...
adjacency_sign_d1  = adjacency(dds_D1_vst_t, power = softPower_d1, type="signed") # this takes a long time.. just wait...


```


## Topological overlap matrix (more fun, please)
```{r}
#=====================================================================================
#
#  Step 2: Turn adjacency into topological overlap matrix
# Calculation of the topological overlap matrix, (TOM)
# and the corresponding dissimilarity, from a given adjacency matrix.
#=====================================================================================

# these matrices take a whiiile, with 13k genes each of the TOM runs take ~ 10 minutes 
# this is why we will save the data here to avoid rerunning again unless we need to
TOM_sign_all       = TOMsimilarity(adjacency_sign_all, TOMType="signed")  # this takes a long time.. just wait...
dissTOM_sign_all   = 1-TOM_sign_all

TOM_sign_d0       = TOMsimilarity(adjacency_sign_d0, TOMType="signed")  # this takes a long time.. just wait...
dissTOM_sign_d0   = 1-TOM_sign_d0

TOM_sign_d1       = TOMsimilarity(adjacency_sign_d1, TOMType="signed")  # this takes a long time.. just wait...
dissTOM_sign_d1   = 1-TOM_sign_d1


# Save checkpoint data - note it took 30 minutes to get this
save(dissTOM_sign_all, dissTOM_sign_d0, dissTOM_sign_d1, # topology overlap matrices  
     file = "Output/NetworkConstruction-postTOM.RData")
```

# CHECKPOINT!!! 
* the previous steps are a drag, so lets load them here and never (hopefully) turn back
```{r}
postTOM_data = load(file = "Output/NetworkConstruction-postTOM.RData") # theres 10 GB of data on this, kinda too much
```

## Clustering clusters
```{r}

sizeGrWindow(12,9)

#=====================================================================================
#
#  Step 3:Call the hierarchical clustering function - plot the tree
#
#=====================================================================================

# Call the hierarchical clustering function
geneTree_sign_all  = hclust(as.dist(dissTOM_sign_all), method = "average");
geneTree_sign_d0   = hclust(as.dist(dissTOM_sign_d0), method = "average");
geneTree_sign_d1   = hclust(as.dist(dissTOM_sign_d1), method = "average");

plot(geneTree_sign_all, # Plot the resulting clustering tree (dendrogram)
     xlab="", 
     sub="", 
     main = "All: Gene clustering on TOM-based dissimilarity - SIGNED",
     labels = FALSE, hang = 0.04);


plot(geneTree_sign_d0, # Plot the resulting clustering tree (dendrogram)
     xlab="", 
     sub="", 
     main = "Day 0: Gene clustering on TOM-based dissimilarity - SIGNED",
     labels = FALSE, hang = 0.04);

plot(geneTree_sign_d1, # Plot the resulting clustering tree (dendrogram)
     xlab="", 
     sub="", 
     main = "Day 1: Gene clustering on TOM-based dissimilarity - SIGNED",
     labels = FALSE, hang = 0.04);
```


## Module identification
```{r}

#=====================================================================================
#
#  Step 4: Set module size and 'cutreeDynamic' to create clusters 
#
#=====================================================================================

# Module identification using dynamic tree cut:

# large modules are favorable, so we set the minimum module size relatively high:
# WGCNA authors recommend diligence when calling module size to avoid too many/too few modules...
minModuleSize = 100; # set this for the subsequent call 



dynamicMods_sign_all = cutreeDynamic(dendro = geneTree_sign_all, 
                                     distM = dissTOM_sign_all,
                                     deepSplit = 1, 
                                     pamRespectsDendro = FALSE,
                                     minClusterSize = minModuleSize);
table(dynamicMods_sign_all) # view the number of genes per module 
#    1    2    3    4    5    6 
# 6766 3424  804  761  498  386


dynamicMods_sign_d0 = cutreeDynamic(dendro = geneTree_sign_d0, 
                                     distM = dissTOM_sign_d0,
                                     deepSplit = 1, 
                                     pamRespectsDendro = FALSE,
                                     minClusterSize = minModuleSize);
table(dynamicMods_sign_d0) # view the number of genes per module 
#    1    2    3    4    5    6 
# 5330 4438  857  769  670  575


dynamicMods_sign_d1 = cutreeDynamic(dendro = geneTree_sign_d1, 
                                     distM = dissTOM_sign_d1,
                                     deepSplit = 1, 
                                     pamRespectsDendro = FALSE,
                                     minClusterSize = minModuleSize);
table(dynamicMods_sign_d1) # view the number of genes per module 
#    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
# 2148 2133 1134 1042 1019  802  750  697  652  551  517  360  275  229  175  155 

```


## Dendrograms
```{r}
#=====================================================================================
#
#  Step 5: convert numeric network to colors and plot the dendrogram
#
#=====================================================================================

# Convert numeric lables into colors

# all
dynamicColors_sign_all = labels2colors(dynamicMods_sign_all) # add colors to module labels (previously numbers)
table(dynamicColors_sign_all) # lets look at this table...
     # blue     brown     green       red turquoise    yellow 
     # 3424       804       498       386      6766       761 

# day 0
dynamicColors_sign_d0 = labels2colors(dynamicMods_sign_d0) # add colors to module labels (previously numbers)
table(dynamicColors_sign_d0) # lets look at this table...
     # blue     brown     green       red turquoise    yellow 
     # 4438       857       670       575      5330       769


# day 1
dynamicColors_sign_d1 = labels2colors(dynamicMods_sign_d1) # add colors to module labels (previously numbers)
table(dynamicColors_sign_d1) # lets look at this table...
      #  black         blue        brown         cyan        green  greenyellow    lightcyan      magenta midnightblue         pink 
      #    750         2133         1134          229         1019          517          155          652          175          697 
      # purple          red       salmon          tan    turquoise       yellow 
      #    551          802          275          360         2148         1042 




# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)

# all
png("Output/All/All_Dendrogram.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign_all, dynamicColors_sign_all, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "All data: Gene dendrogram and module colors 'SIGNED'")
dev.off()



# day 0 
png("Output/Day0/Day0_Dendrogram.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign_d0, dynamicColors_sign_d0, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Day 0: Gene dendrogram and module colors 'SIGNED'")
dev.off()


# day 1
png("Output/Day1/Day1_Dendrogram.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign_d1, dynamicColors_sign_d1, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Day 1: Gene dendrogram and module colors 'SIGNED'")
dev.off()

```


## Eigengenes (now we are getting somewhere!) & inital plot
```{r}
#=====================================================================================
#
#  Step 6: Calculate Eigengenes - view thier connectivity based on 'MEDiss = 1-cor(MEs)'
#
#=====================================================================================

# Calculate eigengenes ============== #

# all
MEList_all = moduleEigengenes(dds_all_vst_t_trim, colors = dynamicColors_sign_all)
MEs_all    = MEList_all$eigengenes 

# day 0
MEList_d0 = moduleEigengenes(dds_D0_vst_t, colors = dynamicColors_sign_d0)
MEs_d0    = MEList_d0$eigengenes 

# day 1
MEList_d1 = moduleEigengenes(dds_D1_vst_t, colors = dynamicColors_sign_d1)
MEs_d1    = MEList_d1$eigengenes 


# Calculate dissimilarity of module eigengenes ============== #

# all
MEDiss_all = 1-cor(MEs_all);
METree_all = hclust(as.dist(MEDiss_all), method = "average") # Cluster module eigengenes

# day 0 
MEDiss_d0 = 1-cor(MEs_d0);
METree_d0 = hclust(as.dist(MEDiss_d0), method = "average") # Cluster module eigengenes

# day 1
MEDiss_d1 = 1-cor(MEs_d1);
METree_d1 = hclust(as.dist(MEDiss_d1), method = "average") # Cluster module eigengenes

# Plot the result ============================================ #
sizeGrWindow(7, 6)

# all 
png("Output/All/All_ClusterEigengenes.png", 1000, 1000, pointsize=20)
plot(METree_all, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
dev.off()

# day 0 
png("Output/Day0/Day0_ClusterEigengenes.png", 1000, 1000, pointsize=20)
plot(METree_d0, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
dev.off()

# day 1
png("Output/Day1/Day1_ClusterEigengenes.png", 1000, 1000, pointsize=20)
plot(METree_d1, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
dev.off()

```


## Cut line to merge eigengene network modules (based on plots in previous chunk)
```{r}

#=====================================================================================
#
#  Step 7: Specify the cut line for the dendrogram (module) - Calc MODULE EIGENGENES (mergeMEs)
#
#=====================================================================================

# all
# not necessary for this dataset - METree above looks to be parsed well into the 6 modules 


# day 0 
# not necessary for this dataset - METree above looks to be parsed well into the 6 modules 


# day 1 - we have many modules and ahould merge these

MEDissThres = 0.4 #  height threshold only nerge modules below 
plot(METree_d1, xlab = "", sub = "")# view what we are merging
abline(h=MEDissThres, col = "red") # add the line

merge_d1 <- mergeCloseModules(dds_D1_vst_t, dynamicColors_sign_d1, cutHeight = MEDissThres, verbose = 3)
mergedColors_d1 = merge_d1$colors # The merged module colors
mergedMEs_d1    = merge_d1$newMEs # Eigengenes of the new merged modules:

```

## Plot new dendrograms
```{r}
#=====================================================================================
#
#  Step 8: Plot dendrogram with the cut line 'MEDissThres' 
#
#=====================================================================================

# use 'mergedColors' we needed to merge related eigengene modules together

# Day 1 - again, only need for any mergingof modules
sizeGrWindow(12, 9)
png("Output/Day1/Day1_ClusterDendrogram_merged.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign_d1, cbind(dynamicColors_sign_d1, mergedColors_d1),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
dev.off()


```


# CHECKPOINT!!! 
# Print Eignegnes 
```{r}
#=====================================================================================
#
#  Step 9: Commit to mergedcolors as 'MEs' and 'moduleColors'
#
#=====================================================================================
colorOrder = c("grey", standardColors(50));

# all
moduleColors_all = dynamicColors_sign_all # moduleColors
moduleLabels_all = match(moduleColors_all, colorOrder)-1 # Construct numerical labels corresponding to the colors
save(MEs_all, dds_all_vst_t_trim, moduleLabels_all, moduleColors_all, file = "Output/All/All-NetworkConstruction_MEdata.RData") # Save 


# day 0
moduleColors_d0 = dynamicColors_sign_d0 # moduleColors
moduleLabels_d0 = match(moduleColors_d0, colorOrder)-1 # Construct numerical labels corresponding to the colors
save(MEs_d0, dds_D0_vst_t, moduleLabels_d0, moduleColors_d0, file = "Output/Day0/Day0-NetworkConstruction_MEdata.RData") # Save 


# day 1
moduleColors_d1 = mergedColors_d1 # moduleColors - NOTE! these are MERGED colors from above!
moduleLabels_d1 = match(moduleColors_d1, colorOrder)-1 # Construct numerical labels corresponding to the colors
save(MEs_d1, dds_D1_vst_t, moduleLabels_d1, moduleColors_d1, file = "Output/Day1/Day1-NetworkConstruction_MEdata.RData") # Save 


# write csv
write.csv(MEs_all, file = "Output/All/All.WGCNA_ModulEigengenes.csv")
write.csv(MEs_d0, file = "Output/Day0/Day0.WGCNA_ModulEigengenes.csv")
write.csv(mergedMEs_d1, file = "Output/Day1/Day1.WGCNA_ModulEigengenes.csv")

```


# LOAD MEs
```{r}
#=====================================================================================
#
#  Prepare for  module trait associations - Eigengene calc - trait data as numeric
#
#=====================================================================================
# identify modules that are significantly associated with the measured  traits (here as treatment history)
# Since we already have a summary profile (eigengene) for each module,
# we simply correlate eigengenes with external traits and look for the  significant associations:


MEs_all          <-  read.csv("Output/All/All.WGCNA_ModulEigengenes.csv") # read  eigengene dataset
rownames(MEs_all)<- MEs_all[,1] # make first column into row names
MEs_all          <- MEs_all[,-1] # omit the first column (now inserted as rownames...)
MEs_all          <- orderMEs(MEs_all) # reorders the columns (colors/modules)

MEs_d0           <-  read.csv("Output/Day0/Day0.WGCNA_ModulEigengenes.csv")
rownames(MEs_d0) <- MEs_d0[,1] # make first column into row names
MEs_d0           <- MEs_d0[,-1] # omit the first column (now inserted as rownames...)
MEs_d0           <- orderMEs(MEs_d0) # reorders the columns (colors/modules)

MEs_d1           <-  read.csv("Output/Day1/Day1.WGCNA_ModulEigengenes.csv")
rownames(MEs_d1) <- MEs_d1[,1] # make first column into row names
MEs_d1           <- MEs_d1[,-1] # omit the first column (now inserted as rownames...)
MEs_d1           <- orderMEs(MEs_d1) # reorders the columns (colors/modules)


```


# MODULE-TRAIT ASSOCIATIONS - Treatments
```{r}
#=====================================================================================
#
# Module trait correlations (Treatments!)
#
#=====================================================================================

# Fed vs. Starved ======================= # - NOTE numeric tables from SAMPLETREE chunks

# all
All.Treatment.asnum             <- data.frame(lapply(All_Treatment, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(All_Treatment))
moduleTraitCor_Treatment_all    <-  cor(MEs_all, All.Treatment.asnum, use = "p");
moduleTraitPvalue_Treatment_all <- corPvalueStudent(moduleTraitCor_Treatment_all, nSamples);
moduleTraitPvalue_Treatment_all
#  day 0
d0.Treatment.asnum               <- data.frame(lapply(D0_Treatment, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D0_Treatment))
moduleTraitCor_Treatment_d0      <-  cor(MEs_d0, d0.Treatment.asnum, use = "p");
moduleTraitPvalue_Treatment_d0 <- corPvalueStudent(moduleTraitCor_Treatment_d0, nSamples);
moduleTraitPvalue_Treatment_d0
# day 1
d1.Treatment.asnum               <- data.frame(lapply(D1_Treatment, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D1_Treatment))
moduleTraitCor_Treatment_d1      <-  cor(MEs_d1, d1.Treatment.asnum, use = "p");
moduleTraitPvalue_Treatment_d1 <- corPvalueStudent(moduleTraitCor_Treatment_d1, nSamples);
moduleTraitPvalue_Treatment_d1

# Line ======================= #  - NOTE numeric tables from SAMPLETREE chunks

# all
All.Line.asnum             <- data.frame(lapply(All_Line, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(All_Line))
moduleTraitCor_Line_all    <-  cor(MEs_all, All.Line.asnum, use = "p");
moduleTraitPvalue_Line_all <- corPvalueStudent(moduleTraitCor_Line_all, nSamples);
moduleTraitPvalue_Line_all
#  day 0
d0.Line.asnum               <- data.frame(lapply(D0_Line, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D0_Line))
moduleTraitCor_Line_d0      <-  cor(MEs_d0, d0.Line.asnum, use = "p");
moduleTraitPvalue_Line_d0 <- corPvalueStudent(moduleTraitCor_Line_d0, nSamples);
moduleTraitPvalue_Line_d0
# day 1
d1.Line.asnum               <- data.frame(lapply(D1_Line, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D1_Line))
moduleTraitCor_Line_d1      <-  cor(MEs_d1, d1.Line.asnum, use = "p");
moduleTraitPvalue_Line_d1 <- corPvalueStudent(moduleTraitCor_Line_d1, nSamples);
moduleTraitPvalue_Line_d1

# Group ======================= #  - NOTE numeric tables from SAMPLETREE chunks

# all
All.Group.asnum             <- data.frame(lapply(All_Group, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(All_Group))
moduleTraitCor_Group_all    <-  cor(MEs_all, All.Group.asnum, use = "p");
moduleTraitPvalue_Group_all <- corPvalueStudent(moduleTraitCor_Group_all, nSamples);
moduleTraitPvalue_Group_all
#  day 0
d0.Group.asnum               <- data.frame(lapply(D0_Group, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D0_Group))
moduleTraitCor_Group_d0      <-  cor(MEs_d0, d0.Group.asnum, use = "p");
moduleTraitPvalue_Group_d0 <- corPvalueStudent(moduleTraitCor_Group_d0, nSamples);
moduleTraitPvalue_Group_d0
# day 1
d1.Group.asnum               <- data.frame(lapply(D1_Group, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(D1_Group))
moduleTraitCor_Group_d1      <-  cor(MEs_d1, d1.Group.asnum, use = "p");
moduleTraitPvalue_Group_d1 <- corPvalueStudent(moduleTraitCor_Group_d1, nSamples);
moduleTraitPvalue_Group_d1


#

```

# MODULE-TRAIT ASSOCIATIONS - Physiology bins
```{r}
#=====================================================================================
#
# Module trait correlations (Physiology bins!)
#
#=====================================================================================

View(Metadata_D0)
Metadata_all

ShellLength_bins
IR_DOM_bins

# Shell Length ================= #
# all 
All_ShellLength  <-  Metadata_all %>% dplyr::select('ShellLength_bins') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate("80-89.2" = as.factor(as.numeric(ShellLength_bins == 1)))  %>%
  dplyr::mutate("89.2-90" = as.factor(as.numeric(ShellLength_bins == 2))) %>% 
  dplyr::mutate(">90" = as.factor(as.numeric(ShellLength_bins == 3))) %>%
  dplyr::select(-ShellLength_bins)
All.ShellLength.asnum             <- data.frame(lapply(All_ShellLength, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(All_ShellLength))
moduleTraitCor_ShellLength_all    <-  cor(MEs_all, All.ShellLength.asnum, use = "p");
moduleTraitPvalue_ShellLength_all <- corPvalueStudent(moduleTraitCor_ShellLength_all, nSamples);
moduleTraitPvalue_ShellLength_all

# day 0 
d0_ShellLength  <-  Metadata_D0 %>% dplyr::select('ShellLength_bins') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate("80-89.2" = as.factor(as.numeric(ShellLength_bins == 1)))  %>%
  dplyr::mutate("89.2-90" = as.factor(as.numeric(ShellLength_bins == 2))) %>% 
  dplyr::mutate(">90" = as.factor(as.numeric(ShellLength_bins == 3))) %>%
  dplyr::select(-ShellLength_bins)
d0.ShellLength.asnum             <- data.frame(lapply(d0_ShellLength, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(d0_ShellLength))
moduleTraitCor_ShellLength_d0    <-  cor(MEs_d0, d0.ShellLength.asnum, use = "p");
moduleTraitPvalue_ShellLength_d0 <- corPvalueStudent(moduleTraitCor_ShellLength_d0, nSamples);
moduleTraitPvalue_ShellLength_d0


# day 1 
d1_ShellLength  <-  Metadata_D1 %>% dplyr::select('ShellLength_bins') %>% # primary Line as Ambient (A) vs. Moderate (M)
  dplyr::mutate("80-89.2" = as.factor(as.numeric(ShellLength_bins == 1)))  %>%
  dplyr::mutate("89.2-90" = as.factor(as.numeric(ShellLength_bins == 2))) %>% 
  dplyr::mutate(">90" = as.factor(as.numeric(ShellLength_bins == 3))) %>%
  dplyr::select(-ShellLength_bins)
d1.ShellLength.asnum             <- data.frame(lapply(d1_ShellLength, function(x) as.numeric(as.character(x))),
                                         check.names=F, row.names = row.names(d1_ShellLength))
moduleTraitCor_ShellLength_d1    <-  cor(MEs_d1, d1.ShellLength.asnum, use = "p");
moduleTraitPvalue_ShellLength_d1 <- corPvalueStudent(moduleTraitCor_ShellLength_d1, nSamples);
moduleTraitPvalue_ShellLength_d1



```


# BRING ON THE HEATMAPS!

# HEATMAPS GROUP
```{r}
#=====================================================================================
#
# Heatmaps for treatment data 
#
#=====================================================================================
sizeGrWindow(10,10)


# Group ==========================================# 

# all 
All_Group.matrix <-  paste(signif(moduleTraitCor_Group_all, 2), "\n(",
                           signif(moduleTraitPvalue_Group_all, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/All/heatmaps/Groups_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_Group_all,
               xLabels = colnames(moduleTraitPvalue_Group_all),
               yLabels = names(MEs_all),
               ySymbols = names(MEs_all),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = All_Group.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-0.6,0.6),
               main = paste("All data: Module-trait relationships - GROUP"))
dev.off()

All.Treatment.text <-  as.matrix(signif(moduleTraitPvalue_Group_all, 3))
All.Treatment.cor  <-  as.matrix(signif(moduleTraitCor_Group_all, 3))
pa                  = cluster::pam(All.Treatment.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/All/heatmaps/Groups_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_Group_all, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "All  WGCNA - group", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 3,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", All.Treatment.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()


# day 0 
d0_Group.matrix <-  paste(signif(moduleTraitCor_Group_d0, 2), "\n(",
                           signif(moduleTraitPvalue_Group_d0, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/Day0/heatmaps/Groups_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_Group_d0,
               xLabels = colnames(moduleTraitPvalue_Group_d0),
               yLabels = names(MEs_d0),
               ySymbols = names(MEs_d0),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = d0_Group.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-1.6,0.6),
               main = paste("d0 data: Module-trait relationships - GROUP"))
dev.off()

d0.Treatment.text <-  as.matrix(signif(moduleTraitPvalue_Group_d0, 3))
d0.Treatment.cor  <-  as.matrix(signif(moduleTraitCor_Group_d0, 3))
pa                  = cluster::pam(d0.Treatment.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/Day0/heatmaps/Groups_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_Group_d0, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 0  WGCNA - group", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        row_km = 4, 
        column_km = 3,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d0.Treatment.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()




# day 1
d1_Group.matrix <-  paste(signif(moduleTraitCor_Group_d1, 2), "\n(",
                           signif(moduleTraitPvalue_Group_d1, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/Day1/heatmaps/Groups_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_Group_d1,
               xLabels = colnames(moduleTraitPvalue_Group_d1),
               yLabels = names(MEs_d1),
               ySymbols = names(MEs_d1),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = d1_Group.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-1.6,0.6),
               main = paste("d1 data: Module-trait relationships - GROUP"))
dev.off()

d1.Treatment.text <-  as.matrix(signif(moduleTraitPvalue_Group_d1, 3))
d1.Treatment.cor  <-  as.matrix(signif(moduleTraitCor_Group_d1, 3))
pa                  = cluster::pam(d1.Treatment.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/Day1/heatmaps/Groups_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_Group_d1, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 1  WGCNA - group", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 3,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d1.Treatment.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()

```

# HEATMAPS SHELLLENGTH
```{r}
#=====================================================================================
#
# Heatmaps for treatment data 
#
#=====================================================================================
sizeGrWindow(10,10)


# ShellLength ==========================================# 

# all 
All_ShellLength.matrix <-  paste(signif(moduleTraitCor_ShellLength_all, 2), "\n(",
                           signif(moduleTraitPvalue_ShellLength_all, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/All/heatmaps/ShellLengths_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_ShellLength_all,
               xLabels = colnames(moduleTraitPvalue_ShellLength_all),
               yLabels = names(MEs_all),
               ySymbols = names(MEs_all),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = All_ShellLength.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-0.6,0.6),
               main = paste("All data: Module-trait relationships - ShellLength"))
dev.off()

All.ShellLength.text <-  as.matrix(signif(moduleTraitPvalue_ShellLength_all, 3))
All.ShellLength.cor  <-  as.matrix(signif(moduleTraitCor_ShellLength_all, 3))
pa                  = cluster::pam(All.ShellLength.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/All/heatmaps/ShellLengths_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_ShellLength_all, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "All  WGCNA - Shell length", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 2,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", All.ShellLength.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()





# day 0 
d0_ShellLength.matrix <-  paste(signif(moduleTraitCor_ShellLength_d0, 2), "\n(",
                           signif(moduleTraitPvalue_ShellLength_d0, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/Day0/heatmaps/ShellLengths_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_ShellLength_d0,
               xLabels = colnames(moduleTraitPvalue_ShellLength_d0),
               yLabels = names(MEs_d0),
               ySymbols = names(MEs_d0),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = d0_ShellLength.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-1.6,0.6),
               main = paste("d0 data: Module-trait relationships - ShellLength"))
dev.off()

d0.Treatment.text <-  as.matrix(signif(moduleTraitPvalue_ShellLength_d0, 3))
d0.Treatment.cor  <-  as.matrix(signif(moduleTraitCor_ShellLength_d0, 3))
pa                  = cluster::pam(d0.Treatment.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/Day0/heatmaps/ShellLengths_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_ShellLength_d0, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 0  WGCNA - ShellLength", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
       #row_km = 4, 
        column_km = 2,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d0.Treatment.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()






# day 1
d1_ShellLength.matrix <-  paste(signif(moduleTraitCor_ShellLength_d1, 2), "\n(",
                           signif(moduleTraitPvalue_ShellLength_d1, 1), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
png("Output/Day1/heatmaps/ShellLengths_heatmap.png", 800, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_ShellLength_d1,
               xLabels = colnames(moduleTraitPvalue_ShellLength_d1),
               yLabels = names(MEs_d1),
               ySymbols = names(MEs_d1),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = d1_ShellLength.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-1.6,0.6),
               main = paste("d1 data: Module-trait relationships - ShellLength"))
dev.off()

d1.Treatment.text <-  as.matrix(signif(moduleTraitPvalue_ShellLength_d1, 3))
d1.Treatment.cor  <-  as.matrix(signif(moduleTraitCor_ShellLength_d1, 3))
pa                  = cluster::pam(d1.Treatment.cor, k = 3)
col_fun             = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("Output/Day1/heatmaps/ShellLengths_heatmap.pdf", width=8, height=6)
Heatmap(moduleTraitCor_ShellLength_d1, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 1  WGCNA - ShellLength", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 2,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d1.Treatment.text[i, j]), x, y, gp = gpar(fontsize = 8))
        })
dev.off()

```




#=====================================================================================
#
# Module eigengene -  MEs boxplots by treatment group
#
#=====================================================================================
MEs_table             <- MEs # new table for plotting 
MEs_table$Sample.Name <- rownames(MEs) # call rows as coolumn to merge with treatment data
MEsPlotting           <- merge(d2.Treatment.data, MEs_table, by = 'Sample.Name') # merge
MEsPlotting_melt      <- melt(MEsPlotting, id=c('Sample.Name', 'Temperature', 'pCO2', 'Salinity', 'pCO2_Salinity', 'All_treatment'))

#plot it - salinity and temperature
png("Output/WGCNA/day2_larvae/Day2_ME_Boxplot_Salinity_Temperature.png", 600, 1000, pointsize=20)
ggplot(MEsPlotting_melt, aes(x=Salinity, y=value, fill = factor(Temperature), shape=Temperature)) +
  geom_boxplot(aes(middle = mean(value)), position=position_dodge(0.8), outlier.size = 0, alpha = 0.5) + 
          stat_summary(fun.y = mean, color = "black", position = position_dodge(0.75),
                       geom = "point", shape = 19, size = 3,
                       show.legend = FALSE) +
          ylab("ModuleEigengene") +
          ylim(-0.5,0.5) +
          scale_fill_manual(values=c("#D55E00","#56B4E9")) +
          geom_hline(yintercept=0, linetype='dotted', col = 'black', size = 1)+
          theme_bw() +
          #theme(legend.position = "none") +
          facet_wrap(~variable)
dev.off()

#plot it - salinity and pCO2
png("Output/WGCNA/day2_larvae/Day2_ME_Boxplot_Salinity_pCO2.png", 600, 1000, pointsize=20)
ggplot(MEsPlotting_melt, aes(x=Salinity, y=value, fill = factor(pCO2), shape=pCO2)) +
  geom_boxplot(aes(middle = mean(value)), position=position_dodge(0.8), outlier.size = 0, alpha = 0.5) + 
  stat_summary(fun.y = mean, color = "black", position = position_dodge(0.75),
               geom = "point", shape = 19, size = 3,
               show.legend = FALSE) +
  ylab("ModuleEigengene") +
  ylim(-0.5,0.5) +
  scale_fill_manual(values=c("#D55E00","#56B4E9")) +
  geom_hline(yintercept=0, linetype='dotted', col = 'black', size = 1)+
  theme_bw() +
  #theme(legend.position = "none") +
  facet_wrap(~variable)
dev.off()

#plot it - salinity and pCO2
png("Output/WGCNA/day2_larvae/Day2_ME_Boxplot_pCO2_Temperature.png", 600, 1000, pointsize=20)
ggplot(MEsPlotting_melt, aes(x=pCO2, y=value, fill = factor(Temperature), shape=Temperature)) +
  geom_boxplot(aes(middle = mean(value)), position=position_dodge(0.8), outlier.size = 0, alpha = 0.5) + 
  stat_summary(fun.y = mean, color = "black", position = position_dodge(0.75),
               geom = "point", shape = 19, size = 3,
               show.legend = FALSE) +
  ylab("ModuleEigengene") +
  ylim(-0.5,0.5) +
  scale_fill_manual(values=c("#D55E00","#56B4E9")) +
  geom_hline(yintercept=0, linetype='dotted', col = 'black', size = 1)+
  theme_bw() +
  #theme(legend.position = "none") +
  facet_wrap(~variable)
dev.off()

#plot it - all treatment
library(forcats)
png("Output/WGCNA/day2_larvae/Day2_ME_Boxplot_All_Treatment.png", 600, 1000, pointsize=20)
MEsPlotting_melt %>% 
  dplyr::mutate(All_treatment = fct_relevel(All_treatment, 'HHH','LHH','HLH','LLH',
                                                           'HHL','LHL','HLL','LLL')) %>% 
    ggplot(aes(x=All_treatment, y=value, fill = factor(Salinity), shape=Salinity)) +
      geom_boxplot(aes(middle = mean(value)), position=position_dodge(0.8), outlier.size = 0, alpha = 0.5) + 
      stat_summary(fun.y = mean, color = "black", position = position_dodge(0.75),
                   geom = "point", shape = 19, size = 3,
                   show.legend = FALSE) +
      ylab("ModuleEigengene") +
      ylim(-0.5,0.5) +
      scale_fill_manual(values=c("#D55E00","#56B4E9")) +
      geom_hline(yintercept=0, linetype='dotted', col = 'black', size = 1)+
      theme_bw() +
      #theme(legend.position = "none") +
      facet_wrap(~variable)
dev.off()



























```{r}

All_data <- load(file="Output/All/All-NetworkConstruction_MEdata.RData")
All_data
#=====================================================================================
#
# geneModuleMembership - geneTraitSignificance - GSPvalue
#
#=====================================================================================
# Gene relationship to trait and important modules: 
# Gene Significance and Module Membership

# We quantify associations of individual genes with our trait of interest (TAOC)

# names (colors) of the modules
modNames = substring(names(MEs_all), 3) # name all the modules, from 3rd character on (first two are ME)
geneModuleMembership = as.data.frame(cor(dds_all_vst_t_trim, MEs_all, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");

D0_L_fed = as.data.frame(as.numeric(All_Group$D0_L_fed)); # Define variable containing the desired column 
names(D0_L_fed) = "D0_L_fed"
D0_L_fed_geneTraitSignificance = as.data.frame(cor(dds_all_vst_t_trim, D0_L_fed, use = "p"));
D0_L_fed_GSPvalue = as.data.frame(corPvalueStudent(as.matrix(D0_L_fed_geneTraitSignificance), nSamples));
names(D0_L_fed_geneTraitSignificance) = paste("GS.", names(D0_L_fed), sep=""); # MA_geneTraitSignificance - pearsons correlation between reads and the MA grop
names(D0_L_fed_GSPvalue) = paste("p.GS.", names(D0_L_fed), sep=""); # corPvalueStudent

#=====================================================================================
#
#  Call annotation data to get module gene data (prep for downstream GO)
#
#=====================================================================================
annot = read.csv(file = "Data/Seq_Reference_Master.csv",header = T) %>% 
  dplyr::select(!Cvirginica_TranscriptID) %>% 
  unique()
dim(annot) # 33777    5
names(annot) # view the column names to call 
probes = names(as.data.frame(dds_all_vst_t_trim)) # get the names of all genes in our dds dataframe 
probes2annot = match(probes, annot$Cvirginica_GeneID) # get the row numbers that overlap with the annotation file to map gene info
# The following is the number genes or 'probes' without annotation:
sum(is.na(probes2annot)) # 595 - these are LOC.. IDs that did not overlap!
# Should return 0.

#=====================================================================================
#
#  BUILD GENE INFO DATAFRAMES
#
#=====================================================================================
# Create the starting data frame
# D0_L_fed_geneTraitSignificance, D0_L_fed_GSPvalue, and moduleColors_all are asociated with dds_all_vst_t_trim
# that was just mapped via 'match to the 'annot$Cvirginica_GeneID', thus the annot data merged based on row number in 'probes2annot'
# and we only need to call the data that is directly associated with dds_all_vst_t_trim, listed above 
geneInfo_GROUPS = data.frame(geneSymbol       = annot$Cvirginica_GeneID[probes2annot],# mapped with 'probes' to correct row numb
                             moduleColor      = moduleColors_all, # associated with dds_all_vst_t_trim
                             KEGG_ID          = annot$Cgigas_KEGGID[probes2annot],# mapped with 'probes' to correct row numb
                             Protein_name     = annot$Cvirginica_Protein_name[probes2annot],# mapped with 'probes' to correct row numb
                             gene_length      = annot$Cvirginica_length[probes2annot],# mapped with 'probes' to correct row numb
                             GO.terms         = annot$Annotation_GO_ID[probes2annot],# mapped with 'probes' to correct row numb
                             D0_L_fed_geneTraitSignificance, D0_L_fed_GSPvalue) # associated with dds_all_vst_t_trim
# call this specific to the module and trait of interest
View(geneInfo_GROUPS)
modOrder = order(-abs(cor(MEs_all, D0_L_fed, use = "p"))); # order by the strength of the correlation between module and trait values for each sample

for (mod in 1:ncol(geneModuleMembership)) { # Add module membership information in the chosen order

  oldNames = names(geneInfo_GROUPS)
  geneInfo_GROUPS = data.frame(geneInfo_GROUPS, geneModuleMembership[, modOrder[mod]], 
                               MMPvalue[, modOrder[mod]]);
  names(geneInfo_GROUPS) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                             paste("p.MM.", modNames[modOrder[mod]], sep=""))
}
# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
geneOrder = order(geneInfo_GROUPS$moduleColor, -abs(geneInfo_GROUPS$GS.D0_L_fed));
geneInfo_GROUPS = geneInfo_GROUPS[geneOrder, ]
View(geneInfo_GROUPS)


#=====================================================================================
#
#  Write csv for the modules and corresponding raw read counts
#
#=====================================================================================
# call the module of interest for follow-up GO analysis 

write.csv(geneInfo_GROUPS, file = "Output/All/All.WGCNA_ModulMembership.csv")

```

#===================================================================================== 
# 
# EXPLORE THE Expression of each module (for loop plots!) BY TREATMENT
#
# evidence from the heatmaps and the regression shwo strong module membership and eigenenge cor strength with lighcyan 
# in response to treatment (primary specifically) and Total Antioxidant capacity 
#
#=====================================================================================
# load necessary data 
load("Output/WGCNA/day2_larvae/day2-networkConstruction-stepByStep.RData") # load dds.d2_vst (in addition to other datasets we do not need...) 

day2_ModuleMembership  <- read.csv(file="Output/WGCNA/day2_larvae/d2.WGCNA_ModulMembership.csv", sep=',', header=TRUE) 

d2.Treatment.data <- read.csv(file="Data/TagSeq/day2.exp.data.csv", sep=',', header=TRUE) %>%  
  dplyr::mutate_if(is.character, as.factor) %>% 
  dplyr::rename('Sample.Name' = 'SapleName_readmatrix') %>% 
  dplyr::rename('pCO2' = 'OA') %>% 
  dplyr::select(c('Sample.Name','Temperature','pCO2','Salinity')) %>% 
  dplyr::mutate(All_treatment = paste( (substr(Temperature,1,1)), 
                                       (substr(pCO2,1,1)), 
                                       (substr(Salinity,1,1)), sep = '')) %>% 
  dplyr::mutate(pCO2_Salinity = substr(All_treatment, 2,3)) # experiment treatment data

# cut the samples we do not need from dds.d2 for rlog transofrmation
dim(d2.Treatment.data) # REMEBER! sample tree cut some of these samples out (2 here!) so we need to do the same beofre plotting 
length(rownames(dds.d2_vst)) # 22 total samples


dds.d2_rlogtrans <- as.data.frame(rlogTransformation(assay(dds.d2))) # rlog transoform the expression data matrix (dds object)
dim(dds.d2_rlogtrans) # 4820   24 - note there are 24 samples here, not ommitted from when run WGCNA
dds.d2_rlogtrans <- dds.d2_rlogtrans[,rownames(dds.d2_vst)]
dim(dds.d2_rlogtrans) # 4820   22 - we now have the correct number of samples!
dds.d2_rlogtrans <- tibble::rownames_to_column(dds.d2_rlogtrans,"TranscriptID") # rownames as first column

# write out this rlog tranformed master data (used for plotting below1) 
write.csv(dds.d2_rlogtrans, file = "Output/WGCNA/day2_larvae/d2_rlog_transformed.csv")

# call the module colors 
modcolor <- as.data.frame(unique(day2_ModuleMembership$moduleColor))
names(modcolor)[1] <- "color"


for(i in 1:nrow(modcolor)) {
  
  # vst read count date - narrow the columns - reshape and rename
  Mod_geneIDs     <- day2_ModuleMembership %>% dplyr::filter(moduleColor %in% modcolor[i,]) %>%  dplyr::select("TranscriptID") %>%  na.omit()
  d2_rlog_Mod      <- dds.d2_rlogtrans %>% dplyr::filter(TranscriptID %in% Mod_geneIDs[,1])
  d2_rlog_Mod_MELT <- melt(d2_rlog_Mod, id=("TranscriptID")) # melt using reshape2
  names(d2_rlog_Mod_MELT)[(2:3)] <-  c('Sample.Name', 'rlog_Expression') # change column names
  
  # merge by common row values 'Sample.Name'
  merged_Expdata_Mod <- merge(d2_rlog_Mod_MELT, d2.Treatment.data, by ='Sample.Name')
  
  # mean Exp response table 
  meanEXp_Mod <- merged_Expdata_Mod %>% 
    select(c('Sample.Name','rlog_Expression','Temperature', 'Salinity', 'pCO2')) %>% 
    group_by(Sample.Name, Temperature, Salinity, pCO2) %>%
    dplyr::summarize(mean.rlogExp = mean(rlog_Expression), 
                     sd.rlogtExp = sd(rlog_Expression),
                     na.rm=TRUE)
  
   
  # summarize datasets further by treatment period  =========================================================================================== #
  # remember:this is a mean of a mean!! First we complete mean vst exp by sample id (compiling all red module genes) - next all sample IDs by the treatment period (below
  # I will use these for mean SE plots 
  
  # Temperature ========================== #
  
  meanEXp_Summary.Temperature <- meanEXp_Mod %>% 
    group_by(Temperature) %>%
    dplyr::summarize(mean = mean(mean.rlogExp), 
                     sd = sd(sd.rlogtExp),
                     n = n(), 
                     se = sd/sqrt(n))
  
  # Salinity treatment ========================== #
  
  meanEXp_Summary.Salinity <- meanEXp_Mod %>% 
    group_by(Salinity) %>%
    dplyr::summarize(mean = mean(mean.rlogExp), 
                     sd = sd(sd.rlogtExp),
                     n = n(), 
                     se = sd/sqrt(n))
  
  # pCO2 treatment ========================== #
  
  meanEXp_Summary.pCO2 <- meanEXp_Mod %>% 
    group_by(pCO2) %>%
    dplyr::summarize(mean = mean(mean.rlogExp), 
                     sd = sd(sd.rlogtExp),
                     n = n(), 
                     se = sd/sqrt(n))
  
  # Salinity treatment ========================== #
  
  meanEXp_Summary.All.Treatment <- meanEXp_Mod %>% 
    group_by(Salinity, Temperature, pCO2) %>%
    dplyr::summarize(mean = mean(mean.rlogExp), 
                     sd = sd(sd.rlogtExp),
                     n = n(), 
                     se = sd/sqrt(n))
  
  # PLOT =========================================================================================== #
  # The errorbars overlapped, so use position_dodge to move them horizontally
  pd <- position_dodge(0.3) # move them .05 to the left and right
   
  # Temperature mean sd plot ========================== #
  
  min_p1 <- min(meanEXp_Summary.Temperature$mean) - max(meanEXp_Summary.Temperature$se)
  max_p1 <- max(meanEXp_Summary.Temperature$mean) + max(meanEXp_Summary.Temperature$se)
  
  Temperature.rlog.Mod <- meanEXp_Summary.Temperature %>% 
    dplyr::mutate(Temperature    = forcats::fct_relevel(Temperature, 'Low', 'High')) %>%
      ggplot(aes(x=Temperature, y=mean, fill=Temperature)) +  # , colour=supp, group=supp))
      theme_classic() +
      geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
      geom_line(position=pd) +
      geom_point(position=pd, size = 4, shape=21) +            
      xlab("Temperature") +
      ylab("rlog gene expression") +                 # note the mean was first by sample ID THEN by treatment
      scale_fill_manual(values=c("grey", "grey")) +
      # scale_color_manual(values=c("#56B4E9","#E69F00")) +
      # ggtitle(paste("Day 7 WGCNA", modcolor[i,], "Module VST GeneExp", sep =' ')) +
      # expand_limits(y=0) +                                                    # Expand y range
      scale_y_continuous(limits=c((min_p1), (max_p1))) +
      theme(text = element_text(size=10), legend.position="none")
  
  
  # Salinity mean sd plot ========================== #
  
  min_p2 <- min(meanEXp_Summary.Salinity$mean) - max(meanEXp_Summary.Salinity$se)
  max_p2 <- max(meanEXp_Summary.Salinity$mean) + max(meanEXp_Summary.Salinity$se)
  
  Salinity.rlog.Mod <- meanEXp_Summary.Salinity %>% 
    dplyr::mutate(Salinity    = forcats::fct_relevel(Salinity, 'Low', 'High')) %>%
      ggplot(aes(x=Salinity, y=mean, fill=Salinity)) +
      theme_classic() +
      geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
      geom_line(position=pd) +
      geom_point(position=pd, size = 4, shape=21) +            
      xlab("Salinity") +
      ylab(NULL) +                 # note the mean was first by sample ID THEN by treatment
      # ylab(paste(modcolor[i,]," Module rlog Gene Expression (Mean +/- SE)", sep = ' ')) +                 # note the mean was first by sample ID THEN by treatment
      scale_fill_manual(values=c("grey", "grey")) +
      # scale_color_manual(values=c("#56B4E9","#E69F00")) +
      # ggtitle("Day 21 WGCNA red' Module VST GeneExp") +
      # expand_limits(y=0) +                                                    # Expand y range
      scale_y_continuous(limits=c((min_p2), (max_p2))) +
      theme(text = element_text(size=10), legend.position="none")
  
  
  # pCO2 mean sd plot ========================== #
  
  min_p3 <- min(meanEXp_Summary.pCO2$mean) - max(meanEXp_Summary.pCO2$se)
  max_p3 <- max(meanEXp_Summary.pCO2$mean) + max(meanEXp_Summary.pCO2$se)
  
  pCO2.rlog.Mod <- meanEXp_Summary.pCO2 %>% 
    dplyr::mutate(pCO2    = forcats::fct_relevel(pCO2, 'Low', 'High')) %>%
        ggplot(aes(x=pCO2, y=mean, fill=pCO2)) +
        theme_classic() +
        geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
        geom_line(position=pd) +
        geom_point(position=pd, size = 4, shape=21) +            
        xlab("pCO2") +
        ylab(NULL) +                 # note the mean was first by sample ID THEN by treatment
        # ylab(paste(modcolor[i,]," Module rlog Gene Expression (Mean +/- SE)", sep = ' ')) +                 # note the mean was first by sample ID THEN by treatment
        scale_fill_manual(values=c("grey", "grey")) +
        # scale_color_manual(values=c("#56B4E9","#E69F00")) +
        # ggtitle("Day 21 WGCNA red' Module VST GeneExp") +
        # expand_limits(y=0) +                                                    # Expand y range
        scale_y_continuous(limits=c((min_p3), (max_p3))) +
        theme(text = element_text(size=10), legend.position="none")
  
  
  
  
  
  # Assemble these together =========================================================================================== #
  #library(ggpubr)
  single.factor.plot <-  ggarrange(Temperature.rlog.Mod, Salinity.rlog.Mod,  pCO2.rlog.Mod,      
                                    plotlist = NULL,
                                    ncol = 3,
                                    nrow = 1,
                                    labels = NULL)
                          
  
  # Summary plot of all treatments ==================================================================================== #
  # All.Treatment mean sd plot
  min_p4 <- min(meanEXp_Summary.All.Treatment$mean) - max(meanEXp_Summary.All.Treatment$se)
  max_p4 <- max(meanEXp_Summary.All.Treatment$mean) + max(meanEXp_Summary.All.Treatment$se)
  
  AllTreatment.rlog.Mod <- meanEXp_Summary.All.Treatment %>% 
    dplyr::mutate(Salinity    = forcats::fct_relevel(Salinity, 'Low', 'High')) %>%
    dplyr::mutate(pCO2        = forcats::fct_relevel(pCO2, 'Low', 'High')) %>%
    dplyr::mutate(Temperature = forcats::fct_relevel(Temperature, 'Low', 'High')) %>%
        ggplot(aes(x=pCO2, y=mean, group=Temperature)) + # group aesthetic connect line (Slaintiy) and color - the x axis in this case is pCO2 
        theme_classic() +
        geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
        geom_line(aes(linetype=Temperature), position=pd) +
        geom_point(position=pd, size = 4, shape=21) +            
        xlab("pCO2") +
        ylab("rlog gene expression") +                 # note the mean was first by sample ID THEN by treatment
         # ylab(paste(modcolor[i,]," Module rlog Gene Expression (Mean +/- SE)", sep = ' ')) +                 # note the mean was first by sample ID THEN by treatment
        scale_fill_manual(values=c("#56B4E9", "#E69F00")) +
        scale_y_continuous(limits=c((min_p4), (max_p4))) +
        theme(text = element_text(size=15)) + 
        facet_wrap(~Salinity) # facetted by temperature
  
  # output   ======================================================================================================== #
  pdf(paste("Output/WGCNA/day2_larvae/ModuleExpression_Treatment/day2_Exp_Module_",modcolor[i,],".pdf", sep = ''), width=9, height=8)
  print(ggarrange(single.factor.plot, AllTreatment.rlog.Mod,         
                  plotlist = NULL,
                  ncol = 1,
                  nrow = 2,
                  labels = NULL))
  dev.off()
  
}
 
